scripts/3.2_assign_primary_facility_work.py

```diff
@@ -0,0 +1,374 @@
+    "assigned_zone"
+)["demand_actual"].transform(lambda x: (x / x.sum()) * 100)
+workzone_assignment_opt["pct_of_d_total_assigned"] = workzone_assignment_opt.groupby(
+    "assigned_zone"
+)["demand_assigned"].transform(lambda x: (x / x.sum()) * 100)
+
+
+# Define the output file path
+os.makedirs(acbm.root_path / "data/processed/", exist_ok=True)
+output_file_path = acbm.root_path / "data/processed/workzone_rmse_results.txt"
+
+# Open the file in write mode
+with open(output_file_path, "w") as file:
+    # (1) RMSE for % of Total Demand
+    predictions = workzone_assignment_opt["pct_of_total_demand_assigned"]
+    targets = workzone_assignment_opt["pct_of_total_demand_actual"]
+
+    rmse_pct_of_total_demand = calculate_rmse(predictions, targets)
+    file.write(f"RMSE for % of Total Demand: {rmse_pct_of_total_demand}\n")
+
+    # (2) RMSE for demand as % of total demand from the same origin
+    predictions = workzone_assignment_opt["pct_of_o_total_assigned"]
+    targets = workzone_assignment_opt["pct_of_o_total_actual"]
+
+    rmse_pct_of_o_total = calculate_rmse(predictions, targets)
+    file.write(
+        f"RMSE for % of Total Demand from the Same Origin: {rmse_pct_of_o_total}\n"
+    )
+
+    # (3) RMSE for demand as % of total demand to each destination
+    predictions = workzone_assignment_opt["pct_of_d_total_assigned"]
+    targets = workzone_assignment_opt["pct_of_d_total_actual"]
+
+    rmse_pct_of_d_total = calculate_rmse(predictions, targets)
+    file.write(
+        f"RMSE for % of Total Demand to Each Destination: {rmse_pct_of_d_total}\n"
+    )
+
+# --- Plots
+
+# Plot the demand_actual and demand_assigned values as a line graph for n origin_zones.
+plot_workzone_assignment_line(
+    assignment_results=workzone_assignment_opt,
+    n=10,
+    selection_type="top",
+    sort_by="actual",
+    save_dir=acbm.root_path / "data/processed/plots/assigning/",
+)
+
+# Plot the demand_actual and demand_assigned values as a heatmap for n origin_zones.
+plot_workzone_assignment_heatmap(
+    workzone_assignment_opt,
+    n=20,
+    selection_type="top",
+    sort_by="assigned",
+    save_dir=acbm.root_path / "data/processed/plots/assigning/",
+)
+
+
+#### ASSIGN TO FACILITY ####
+
+# 1. Get neighboring zones
+
+# Sometimes, an activity can be assigned to a zone, but there are no facilities
+# in the zone that match the activity type. In this case, we can search for matching
+# facilities in neighboring zones.
+
+zone_neighbors = Queen.from_dataframe(boundaries, idVariable="OA21CD").neighbors
+
+# 2. select a facility
+
+# apply the function to a row in activity_chains_ex
+activity_chains_work[["activity_id", "activity_geom"]] = activity_chains_work.apply(
+    lambda row: select_facility(
+        row=row,
+        facilities_gdf=osm_data_gdf,
+        row_destination_zone_col="dzone",
+        row_activity_type_col="dact",
+        gdf_facility_zone_col="OA21CD",
```

Comment:
When I run script 3.2, I seem to have floats for "dzone" so I think it might lead to a comparison that is always false between integer zones and OA21CD [here](https://github.com/Urban-Analytics-Technology-Platform/acbm/blob/bdbed0a2020216541160560d52d4a1f109abe077/src/acbm/assigning/primary_select.py#L206)?

Reply:
dzone is the zone id (in this case OA21CD). If you look at my last two commits here, you'll see that I:
- I've updated the logs to show what's happening. 
- Fixed a bug **in 3.1** [here](https://github.com/Urban-Analytics-Technology-Platform/acbm/pull/35/commits/7c02275adf9e00a8f818a11c834e825bc6b04f6b)

Could you rerun 2, 3.1 and 3.2 and then see the log for `assigning_primary_locations`?

Reply:
Looks good, I've gone ahead and rerun all the scripts successfully now following adding two small changes in https://github.com/Urban-Analytics-Technology-Platform/acbm/pull/35/commits/5cde24aa170c1f4609fd93e49ee0ad03a37044f7 and https://github.com/Urban-Analytics-Technology-Platform/acbm/pull/35/commits/edf51ea18c367ccb936f383380b91a4e4b82b91a